src/colorizer/utils/data_utils.ts

```diff
@@ -397,3 +401,134 @@ export function cloneLabel(label: LabelData): LabelData {
+  // Especially for discontinuous lines, IDs may be repeated. Cache results to
+  // avoid repeat computation.
+  const idToColor = new Map<number, [number, number, number]>();
+  const vertexColors = new Float32Array(ids.length * 3);
+  for (let i = 0; i < ids.length; i++) {
+    const id = ids[i];
+    if (idToColor.has(id)) {
+      vertexColors.set(idToColor.get(id)!, i * 3);
+    } else {
+      const color = computeColorFromId(id, params);
+      const rgb = [color.r, color.g, color.b] as [number, number, number];
+      idToColor.set(id, rgb);
+      vertexColors.set(rgb, i * 3);
+    }
+  }
+  return vertexColors;
+}
+
+/**
+ * Calculates 3D coordinates for points in a track's path, to be rendered as a
+ * 3D polyline (see Three.js `Line2`). Also returns the object IDs for each
+ * vertex, which can be used to color the points based on feature data.
+ * @returns An object with two properties:
+ * - `ids`: An array of object IDs corresponding to each point in the path.
+ * - `points`: A Float32Array of 3D centroids for each point in the track's
+ *   path. Coordinates are given in terms of the centroid data (typically as
+ *   pixels/voxels in the frame). The length will be `3 * ids.length`.
+ */
+export function computeTrackLinePointsAndIds(dataset: Dataset, track: Track): { ids: number[]; points: Float32Array } {
+  if (track.centroids.length === 0) {
+    return { ids: [], points: new Float32Array(0) };
+  }
+  const points = new Float32Array(track.duration() * 3);
+  const ids = [];
+
+  // Tracks may be missing objects for some timepoints, so use the last known
+  // good value as a fallback
+  // TODO: This currently makes continuous lines, even if the track has gaps.
+  // Add an option to visualize discontinuities. (This will also require use
+  // of `LineSegments2` instead of `Line2`.)
+  let lastTrackIndex = 0;
+  for (let i = 0; i < track.duration(); i++) {
+    const absTime = i + track.startTime();
+
+    let trackIndex = track.times.findIndex((t) => t === absTime);
+    if (trackIndex === -1) {
+      // Track has no object for this time, use fallback
+      trackIndex = lastTrackIndex;
+    } else {
+      lastTrackIndex = trackIndex;
+    }
+
+    const centroid = dataset.getCentroid(track.ids[trackIndex])!;
+    points.set(centroid, i * 3);
+    ids.push(track.ids[trackIndex]);
+  }
+  return { ids, points };
+}
+
+/**
+ * Normalizes 3D centroids to 2D canvas space in-place, where the X and Y coordinates are
+ * in the range [-1, 1] and the Z coordinate is always zero.
+ */
+export function normalizePointsTo2dCanvasSpace(points: Float32Array, dataset: Dataset): Float32Array {
+  for (let i = 0; i < points.length / 3; i++) {
+    points[3 * i + 0] = (points[3 * i + 0] / dataset.frameResolution.x) * 2.0 - 1.0;
+    points[3 * i + 1] = -((points[3 * i + 1] / dataset.frameResolution.y) * 2.0 - 1.0);
+    // Z coordinate is always zero for 2D canvas space
+    points[3 * i + 2] = 0;
+  }
```

Comment:
nit: might be worth either breaking `i * 3` into its own variable, or just deciding to increment `i += 3`? rather than multiplying by 3 a bunch of times